### DISCUSSION TITLE
3Q0907530AC A3 8V - Profiles memory after restart FIXED

### DISCUSSION THREAD

**Discussion by Mich795** [jilleb/MQB-FPA#57] [Q&A] [answered]
Hi, I tried to edit dataset for gateway 3Q0907530AC for a3 8v. I tried to fix the problem of storing the profiles after restart, but when I load dataset on car, it doesn't memorize the profiles yet. I am attaching what I did, I hope someone is able to check it (I saw that others have had problems on a3 8v) .. 
Thanks :)

[3Q0907530AC.zip](https://github.com/jilleb/MQB-FPA/files/8609989/3Q0907530AC.zip)

**Comment by G1Apples** [comment]
> Hi, I tried to edit dataset for gateway 3Q0907530AC for a3 8v. I tried to fix the problem of storing the profiles after restart, but when I load dataset on car, it doesn't memorize the profiles yet. I am attaching what I did, I hope someone is able to check it (I saw that others have had problems on a3 8v) .. Thanks :)
> 
> [3Q0907530AC.zip](https://github.com/jilleb/MQB-FPA/files/8609989/3Q0907530AC.zip)

Could you maybe post the unedited dataset ? it would help me alot
(if you got it already working i would also take the finished dataset)

Thanks alot.

**Reply by Mich795** [reply]
Dataset for lear or continental unit?

**Reply by G1Apples** [reply]
It should be the Continental unit "Sw 5388"

**Reply by Mich795** [reply]
Should be this...
[3Q0907530AC_5388.zip](https://github.com/jilleb/MQB-FPA/files/9447072/3Q0907530AC_5388.zip)

**Reply by G1Apples** [reply]
Thank you very much, 
have you solved ur problems or is this still an open question ?
Im going to try it anyways and could send you the finished Dataset if it works.

**Reply by Mich795** [reply]
It's an open question.. not solved

**Comment by olli991** [comment]
[19_3Q0907530AX_only_FPA_A3_8V_Test1_20220911_010346.zip](https://github.com/jilleb/MQB-FPA/files/9541517/19_3Q0907530AX_only_FPA_A3_8V_Test1_20220911_010346.zip)
Try this it will remeber everything in every mode. It is based on your first.zip and the corresponding FPA DS in there. 

You original zip dos not include any profiles that could be remembered.

**Reply by Mich795** [reply]
It doesn’t work.. Leaving first profile, when I turn it back on:

- Efficency > Individual
- Comfort > Comfort
- Auto > Auto
- Dynamic > Dynamic
- Individual > Auto

(Turn off in > Turn on in)

**Reply by olli991** [reply]
I'll take another look at it soon

**Reply by Cuzoe** [reply]
You can give this a try... the only thing I have modified is 0x341h... changed from 2 > 9
- This should fix one of the profiles returning to the incorrect profile, but for reasons (below) I'm not sure which...
[19_3Q0907530Ax_only_FPA_A3_8V_ReturnTest_20220918_064152.zip](https://github.com/jilleb/MQB-FPA/files/9594083/19_3Q0907530Ax_only_FPA_A3_8V_ReturnTest_20220918_064152.zip)

There is some weirdness in this dataset that makes it difficult to decipher...
- This was the first thing I noticed... the order is determined in the 0x7D6 - 0x7F6 range
![image](https://user-images.githubusercontent.com/101527335/190902774-7dab621f-9da9-4b9f-ae4c-84184588a6b8.png)
- In trying to "fix" the above order, I noticed that displayed profiles have been switched around...
  - See here... https://github.com/jilleb/MQB-FPA/wiki/dataset#0x32d---0x339-profiles


![image](https://user-images.githubusercontent.com/101527335/190902958-b226a38d-b88c-4e35-a0ef-3727adce6c19.png)
This makes sense if, for example, your car does not have the FEC for Comfort, and you want to replace Eco.  I don't know if this car has Comfort but I think yes because DCC is configured... so I'm not sure why profiles have switched...
- [4] Offroad profile is being displayed as (5) Eco
- [5] Eco profile is being displayed as (6) Individual
- [6] Individual profile is not being displayed (not set)


My recommendation, to reduce confusion, would be to assign every profile to its "proper" display
- This is from my dataset, I have Comfort FEC
- ![image](https://user-images.githubusercontent.com/101527335/190903376-b63a8f7a-faf5-4630-bea0-e0d1283a3256.png)
- Then replace what you want (for example Eco display as Comfort, or Sport display as Race)


There are still other things that are sort of strange, although they stem from the previously mentioned profile switching...
![image](https://user-images.githubusercontent.com/101527335/190903991-354b6840-7675-4dbb-897a-db092c3a5ef9.png)
- All of the profiles except "Eco" are set to show nothing if press the i-info button on the Drive Mode HMI screen
- This is the 0x4Eb range, named _struct available_choices_at_individual_profile_
  - Typically this will show you that in Profile X, Control (X) is set to (X)... see my profile
  - ![image](https://user-images.githubusercontent.com/101527335/190904273-53ac222a-5c58-42d0-b87d-e0820149bb73.png)
  - This is just informational, I can not change anything when looking at the info page for Comfort, Normal, Sport, Eco
  - But for Individual Profile, the i-info button on the HMI is replaced with "Adjust," so now I have choices... my car below
  - ![image](https://user-images.githubusercontent.com/101527335/190904617-e986ab88-300d-4aff-9b78-b9861825173c.png)
- But on your profile, "Eco" is set to show choices, because it is being displayed as Individual
- ![image](https://user-images.githubusercontent.com/101527335/190904740-2323fda9-ed47-4151-87c5-734fab797344.png)
- I have not tried this but assuming it works, it still makes your profile challenging to decipher

All that to say... If you tell me what FEC's you have (specifically if you have Comfort/Offroad) I can edit/re-organize this profile to a "normal" state when I get some time.   I think it will make it easier for you to make your own custom changes going forward.

**Reply by Mich795** [reply]
Yes I have Comfort profile active, but I do not know that I also have off-road, I don't think so (where can I check with certainty the necessary FECs codes?).
I compared a RS3 dataset (Untiteled2) with mine that you send me back (Untiteled1), everything is the same except what is attached in this photo:
![info1](https://user-images.githubusercontent.com/101910756/190922332-c61e58cc-6f2e-4eb1-a38a-c7ea50c675d8.png)
Maybe that's the problem..

**Reply by Cuzoe** [reply]
I tried downloading your dataset from the post you made on May 3 but was not successful.  Maybe that's a VCP file?
So I modified the file that @olli991 posted on Sep 10th.  If you have your original file I could start from there.

Anyway, in the picture you posted that is looking at the range starting at 0x2D2, which is not correct.  There are 2 sets of values that determine how the car behaves when you turn off then back on...

**Range starting at 0x339h - Driving Modes after restart**
- This range is for entire mode... So if you turn off in Mode (x), what mode when you turn on?
- This can be the same (example i) or it can be different (example ii)
  1) Turn off in Normal, turn back on in Normal... Turn off in Eco, turn back on in Eco (this is what you want)
  2) Or... Turn off in Sport, turn back on in Normal... Turn off in Eco, turn back on in Normal (this is not what you want)
      - Some people might want car always in Normal when they start
 
**Range starting at 0x2D2h - Individual Controls after restart**
- This is range is for each control, _inside_ of each Driving Mode
- I'll give an example from the Golf R...
  - Turn off in Race, turn back on in Race (set in the 0x339h range), _in Race the Exhaust Valves are set to Race also, fully open_
  - But inside of Race... Exhaust Valves do not return to Race (fully open) on restart, they only restart Normal (half open)
  - So this means car turns off in Race, turns back on in Race, but to get exhaust valves fully open you have to use Mode Button to select Race again... then exhaust valves will be in Race (fully open)

@olli991 set all of the controls in 0x2D2 to remember their position, so that is not your problem.
Your problem is in the 0x339h range... but it is difficult to fix because the profiles are all mixed up
- Factory would not have these modifications, someone has made a few mistakes in modification
- I made same mistakes when using the earlier version of the template, but the new template made them easy to see/fix

If you have the factory dataset that would be really helpful.  Of course I (or you) can make a fully custom dataset.  Best to start with a factory dataset that is similar to your car, so you don't have to make so many changes.
- My car came without Driving Modes (which were retrofit-ed), and has several other retrofits so I had no choice, had to make my own dataset (but used a GTI dataset as a starting point)

**Reply by Mich795** [reply]
Oh ok, now everything is clearer… You have to consider that my car is an A3 PFL (which already had the driving profiles in the old gateway unit 5Q0) on which the new gateway 3Q0 of the A3 FL has been retrofitted.  Unfortunately I don't have the original (5Q0 and 3Q0) dataset of my car. I have another one, which is the one I sent in May, maybe I could find a very similar dataset (same Lear gateway) but SVM is needed because of i have only VIN of this car... It would be useful if someone had one clear dataset to try

**Reply by Cuzoe** [reply]
Gotcha... If you have VCDS, or you can get a hold of VCDS, post a current AutoScan of your car.  And maybe describe the full spec of your car.  I would ask only model and trim but I don't know if you are in the US so spec might be different.

It will take a few days since the weekend is ending but I'll make a dataset for you.  It will be a good starting point based on your car equipment.

Obviously you can then modify for yourself.  Or people here can assist when possible.

We don't know your FECs so I will assume no Offroad but I will make one profile with Comfort and one without.

I also had originally 5Q0 in my PFL Golf, 3Q0 retrofit... but I had no driving modes either 😂.

**Reply by Mich795** [reply]
Yes, I have VCDS, I'll do a scan in the next few days...
Anyway, my A3 is a normal A3 8v sedan TDI quattro PFL, european model (so I think without off-road 😅 in profile select I have from factory: Comfort, Auto, Normal, Dynamic, Individual)..

I appreciate, thank you :) 
(when I have a functional dataset I will post it on this page for anyone who has this problem in the future)

**Reply by Cuzoe** [reply]
Got it, DCC also?

**Reply by Mich795** [reply]
> Got it, DCC also?

Nope (maybe a future retrofit.. but not planned), but have soundaktor retrofit done

**Reply by Cuzoe** [reply]
My car has DCC and Soundktor, both retrofit.  In the US, on VW's at least, rare to see comfort on cars with no DCC.

Last question for now... you listed Comfort, Auto, Normal, Dynamic, Individual.
I assume you meant Comfort, **Eco**, Normal, Dynamic, individual?

I should be able to get anything else I need from the AutoScan.

**Reply by Mich795** [reply]
Yes, sorry.. i'm wrong. Those are my profiles

![image](https://user-images.githubusercontent.com/101910756/190930413-704a6f12-3553-43bc-bb05-ad086a9b9d96.jpeg)

![image](https://user-images.githubusercontent.com/101910756/190930418-84ca4ffd-7aca-4b57-b96c-d10b8f34efbe.jpeg)

**Reply by Cuzoe** [reply]
Got it, I've got a little time while the kiddo naps... 6MT or DSG?

And do you want controls grouped or everything separate?

Factory will group many things (Like Engine and Transmission).  I prefer no groups, so I can change anything I want in Individual.

**Reply by Mich795** [reply]
6MT… Everything separated seems more interesting, then it would be useful to understand what to change in dataset to eventually return to the factory settings

**Reply by Cuzoe** [reply]
Okay... and your car had engine sound option for individual?

Generally this is either internal soundaktor, external soundaktor or exhaust valves.  You said you retrofit-ted soundaktor (internal)?  

So does your car also have external and/or exhaust valves?

Edit: may have answered my own question with google.  Looks like you have exhaust valves...

So now your car has internal soundaktor and exhaust valves?

**Reply by Cuzoe** [reply]
Ahhh, so many questions... do you have ACC?  And if yes... do you want ACC control in the driving profiles, or will you use the Vehicle Menu?

Adaptive front lighting?
Start Stop?

I think that's all, haha.

**Reply by Mich795** [reply]
No, I only installed the internal soundaktor (which was not foreseen on my car).. it produces different sounds depending on the profile set (stronger in Dynamic) and in Individual profile I can choose which sound to play of the other profiles... up to now no retrofit of the external one, but if I find the components at a good price maybe I will do it in the future.

About ACC, no ACC.. only standard cruise control (speed setting only), eventually standard settings will be ok

**Reply by Cuzoe** [reply]
This should work... the zip has the "file" for 010 editing and an XML for ODIS (of course you can run the script on the file if you use VCP)
[8V A3 TDI 6MT.zip](https://github.com/jilleb/MQB-FPA/files/9596346/8V.A3.TDI.6MT.zip)

I'm not sure if your car has any ambient lighting.  So the configurations are in the dataset but none of them will display on your HMI.  If you have ambient lighting let me know, that's an easy change.

Every driving mode will remain the same after car off, then back on.  And all the controls in each profile will remain.
I don't know if you care about the driving mode button light (orange) so I left that factory, orange when you are in any mode except Normal.

**Reply by Mich795** [reply]
Oh, cool... Yes I have ambient LEDs (white) in doors and footwell... It would be interesting to add the color change option with Dynamic driving mode from white to red (but I would have to physically change the LED, now it is only white, without the possibility of changing)

I quickly checked  your file, is air conditioning in sport set to eco? is it possible to set it to Normal? (however it's a completely different layout from the original one, interesting)

**Reply by Mich795** [reply]
> This should work... the zip has the "file" for 010 editing and an XML for ODIS (of course you can run the script on the file if you use VCP)
> 
> [8V A3 TDI 6MT.zip](https://github.com/jilleb/MQB-FPA/files/9596346/8V.A3.TDI.6MT.zip)
> 
> 
> 
> I'm not sure if your car has any ambient lighting.  So the configurations are in the dataset but none of them will display on your HMI.  If you have ambient lighting let me know, that's an easy change.
> 
> 
> 
> Every driving mode will remain the same after car off, then back on.  And all the controls in each profile will remain.
> 
> I don't know if you care about the driving mode button light (orange) so I left that factory, orange when you are in any mode except Normal.

![image](https://user-images.githubusercontent.com/101910756/191042475-51df2b76-9d40-426a-8c09-42d94eeeb42f.jpeg)

Update* No Individual profile with this mod

**Reply by Cuzoe** [reply]
Is everything else there?  I have a look in about an hour let's see what I screwed up, haha.  Either I screwed something up or we have discovered why the factory Audi data set did that weird profile switching.  If I can't find my mistake I'll set it up using the switched profiles 😞.

Yes, typically in sport the air conditioner is set to Eco, uses less of your engine power, haha.  But I can change it.

**Comment by Cuzoe** [comment]
As expected, silly mistake on my part.  Should be fixed now.  Also changed HVAC to "Normal" for Dynamic mode, see V2 attached.
[8V A3 TDI 6MT V2.zip](https://github.com/jilleb/MQB-FPA/files/9601201/8V.A3.TDI.6MT.V2.zip)

If you ever decide to go dual color ambient lighting we'll have to experiment.  I'm not sure how that gets configured.  I have RGB ambient lighting in my car so I know how to modify dataset for that.  

Also, please still post the AutoScan when you can.  I can at least check the profile against your full configuration and see if there's something to update/change

**Reply by Mich795** [reply]
Here there is a scan by vcds

Log.txt

**Reply by Cuzoe** [reply]
Okay got it, have you tried the V2 data set? I see Battery Monitoring Module and headlight range module... so a few more questions... 

1) Do you have Start/Stop?
   - If yes I can disable it in one (or all) modes let me know, otherwise you can keep using the button
2) Do you have adaptive front lights?  
   - If yes, this should be included in the profiles also.

Another note: I see you are running 8S0 907 159 on the Soundaktor, I assume from @jilleb Soundaktor GitHub
- The only 8S0 xxx xxx software that has different sound for Comfort/Eco, Normal, Sport is FL_8S0907159___0004_UNLOCKED_2.odx
  - _Comfort and Eco have the same sound_
- The other 8S0 xxx xxx firmware have the same sound in Comfort, Normal and Eco... different sound _only_ in Sport

**Reply by Mich795** [reply]
I haven't tried the V2 yet. Anyway: 
1. yes I have start&stop, maybe it would be interesting if it is switched off in Dynamic (But it's ok also as it is)
3. I have normal xenons lamp, so I assume nothing can be set

In the end, about soundaktor: yes it's 8S0 firmware, but not standard one by jilleb: https://github.com/jilleb/mqb-soundaktor/issues/13 (look at the end)

**Comment by Cuzoe** [comment]
This should turn off Start/Stop in Dynamic mode.  I did not add a setting for this to Individual (because of some weirdness with the naming).  Anyway, it should work.  Start/Stop will be On with all other modes.  I guess make sure the button still works to turn it on/off.  I have no start/stop so never tested.
[8V A3 TDI 6MT V3.zip](https://github.com/jilleb/MQB-FPA/files/9602910/8V.A3.TDI.6MT.V3.zip)

**Reply by Mich795** [reply]
I've tried it... for this also, no Individual profile visualization..

Little update* 
Another guy ( @lito69 ), has produced a mod of my original dataset that seems to work correctly (memory profile ok)... I'm attaching it here, try to check it if you like and let me know :)
FPA_test.txt

**Comment by Cuzoe** [comment]
**I see now what was done.  I'm no longer in the game of editing Audi datasets, haha.  They are messy (to me).**  

0x365 range, Struct Profile [3] is being used for the Eco/Efficiency setting... normally this is Offroad
- Many (VW) cars don't have this Offroad FEC so they will not show a button for this
- Maybe Audi has the FEC, but then the dataset replaced Offroad with Eco
- Dataset is sending the values for Eco, so it works as Eco

0x365 range, Struct Profile [4] is being used for Individual... normally this is Eco
- I didn't know why, but from your tests of my profiles, seems Audi does not have the FEC for Individual
- So then Audi dataset replaced Eco with Individual
- Dataset is sending values for Individual, so it works as Individual

0x365 range, Struct Profile [8] is being used for Offroad... normally this Lift/PHEV 2
- Many cars don't have Lift/PHEV 2 FEC so they will not show a button for this
- Seems your car also does not have this FEC, so this button is not shown
- But still Audi dataset replaced Lift/PHEV 2 with Offroad
- Dataset is sending values for Offroad, so it works as Offroad (if you have the Lift/PHEV 2 FEC)

0x4EB range, Struct available_choices_at_individual_profile [4] is set with options
   - This allows you to change settings in the Individual Profile (which replaced normal Eco, haha)
   - For some reason Audi does not let you see the control modes of other settings, maybe MMI can't do it
   - I don't like this, although it doesn't really matter if you can see them when you can't change them

### So what have I learned?
- 0x339h range, byte profile_returns_after_restart can be misleading _if_ profiles have been switched 
- it refers to the position on the list of profiles (keep in mind Profiles 0-11 = Value 1-12)
- My confusion is that in the output window, it gives Driving Mode names instead of Profile positions
  - Enter Value 4 in Profile [3] and the name will show Offroad, but really it should say Profile 4
  - Enter Value 6 in Profile [1] and the name will show Race, but really it should say Profile 2
- Giving driving mode names works great on VW where the profiles haven't been switched, lol



Of course you can replace anything with anything else, as mentioned in the [Wiki here](https://github.com/jilleb/MQB-FPA/wiki/dataset#0x32d---0x339-profiles)
- I would think most likely use would be have Race show instead of Sport
- But then you need to pay close attention to the 0x339h range, byte profile_returns_after_restart

**Reply by Mich795** [reply]
I tried to change "Dynamic" flag into "Race" one, but without success (it would be interesting to understand if it is possible to replace those two words on Audi too)... 
I must say that I still have a lot to study about it... 

Next hours I will post the complete ready to use dataset done by @lito69, I hope anyone who has the same problem can solve it more easily :)

**Reply by Cuzoe** [reply]
Does your car show flags?  Or just the list like you pictured?

In my car, there is no list, only the screen with flags...  Maybe more logic differences on Audi
![image](https://user-images.githubusercontent.com/101527335/191316909-989b3cf2-0607-4cce-8536-33df90b2f439.png)
Not my screen, actually jilleb's, but you get the point

**Reply by Mich795** [reply]
Profiles, when correctly displayed, look like this:

![1](https://user-images.githubusercontent.com/101910756/191318523-1ce5ee9a-7262-4a7a-8da2-414c3dfef684.jpeg)

When I tried to change "Dynamic" to "Race", the word "Dynamic" disappeared and nothing appeared on its place (displaying only: efficency, comfort, auto and individual). It may be possible that I made a mistake in modifying some value in dataset or that MMI isn't set up for the word "Race"... I currently don't know

**Reply by Cuzoe** [reply]
Gotcha, quite a few differences Audi to VW.  I'm limiting myself now to VW only 😁.

I also don't like having anything in my data set that does not apply to my car.  For example I removed front differential lock and transmission management from my personal data set.

Of course this doesn't matter as long as you don't show the controls for these items on the HMI.  Even if you do show the controls on the HMI they simply won't do anything, but won't cause a problem. I just can't sleep well knowing things are in there that don't belong 🫣.

**Comment by Mich795** [comment, marked as answer]
Here we go, thanks to @lito69 , @Cuzoe and @olli991, the complete dataset ready to be used by anyone who needs it :)

[3Q0907530AC_Audi_A3_8V - Dataset.zip](https://github.com/jilleb/MQB-FPA/files/9609700/3Q0907530AC_Audi_A3_8V.-.Dataset.zip)

**Reply by lito69** [reply]
you only had to place the cluster info in the same order and before calculating in binary the active bits for cluster info, the modes in audi are 0102030507 for audi fecs 1 2 3 4 5 and mode return 1 2 3 4 5 which is the mode mode

**Reply by Cuzoe** [reply]
Although I can see what was done in the data set, I don't follow what you posted here...

Audi modes 0102030507... can you explain?
And Audi FECs 1 2 3 4 5... What are they?

**Reply by lito69** [reply]
yes, i was also wrong when interpreting, all except audi have fecs box 01020304050007 and remember 0102030405ff07, in those cars the installed swap is usually 1000120 1000130 10001500 1000170, those swaps are the box number, not the fec number, that's why in audi that has swap from 1 to 5 it has 0102030507 in the dataset and it works the same as in mine, it uses the first 5 Date Bytes I'm with the translator and I don't know if he understands.Then, below for the first byte the profile info, for the second byte the info and if in byte 4 there is 05 (echo) the info is in zone 4, not in 5. The modes to remember is the byte number, not the fec, that's why in audi 04 it would be the fourth which is 05 (echo). Sorry but I'm with the translator.

**Reply by lito69** [reply]
Audi in 19 swap state has swap for profiles from 1 to 5, and in those bytes there is mode 01 02 03 05 07, the swap enables the box, not the fec itself

**Comment by Cuzoe** [comment]
I think I'm following you...

### VW 
FEC box... 01020304050007
- 01 (Comfort) 02 (Normal) 03 (Sport) 04 (Offroad) 05 (Eco) 00 (nothing, **except R models?**) 07 (Individual)
With memory in 01 02 03 04 05 ff 07... so always returns to selected profile
Normally with swaps 1000120 (Normal) 1000130 (Sport) 1000150 (Eco) 1000170 (Individual)
- Cars with DCC might also have swap 1000110 (Comfort)
- Cars with R Spec (Golf R, Tiguan R, etc)?
   - FEC box 010200040050607, with swap 1000110, 1000120, 1000150, 1000160, 1000170
   - Then VW replace Sport with Race

### AUDI
FEC box... 0102030507
- 01 (Comfort) 02 (Normal) 03 (Sport) 05 (Eco) 07 (Individual)
With memory in 01 02 03 05 07
Normally with swaps 1000110 (Comfort) 1000120 (Normal) 1000130 (Sport) 1000140 (Offroad) 1000150 (Eco)
- Cars even without DCC get comfort... maybe because of Center Diff Lock setting?

**For this reason Audi moves Eco into box 4, and Individual into box 5?**

- Audi configure Offroad in Profile 8 (box 9) but with no FEC/Swap it does not show
- VW configure Offroad in Profile 3 (box 4) but with no FEC/Swap it does not show

**Reply by lito69** [reply]
Vw fec 04 offroad. Everything else you have explained well. some have race fec 6.
The cars that I have seen with fec 01 comfort are because they have dcc

**Reply by Cuzoe** [reply]
Yes, Comfort only with DCC.  I have DCC retrofit, and comfort FEC.

I do wonder though, how jilleb got so many to show on his screen, haha.  He has all the FEC's in the world!
https://www.drive2.com/l/529965430295494961/

**Reply by lito69** [reply]
I wonder too, a long time ago I needed to add snow to an ateca (which disappeared when updating gw) and I didn't know how to do it

**Reply by olli991** [reply]
You can "buy" this if you know the right people but it's rly expensive and not worth it tbh.
